src/main/java/com/wanderersoftherift/wotr/world/level/levelgen/processor/AttachmentProcessor.java

```diff
@@ -0,0 +1,145 @@
+
+    private final BlockState blockState;
+    private final int requiresSides;
+    private final boolean requiresUp;
+    private final boolean requiresDown;
+    private final float rarity;
+    private final StructureRandomType structureRandomType;
+    private final long seed;
+
+    public AttachmentProcessor(BlockState blockState, int requiresSides, boolean requiresUp, boolean requiresDown, float rarity, StructureRandomType structureRandomType, long seed) {
+        this.blockState = blockState;
+        this.requiresSides = requiresSides;
+        this.requiresUp = requiresUp;
+        this.requiresDown = requiresDown;
+        this.rarity = rarity;
+        this.structureRandomType = structureRandomType;
+        this.seed = seed;
+    }
+
+    @Override
+    public List<StructureTemplate.StructureBlockInfo> finalizeProcessing(ServerLevelAccessor serverLevel, BlockPos offset, BlockPos pos, List<StructureTemplate.StructureBlockInfo> originalBlockInfos, List<StructureTemplate.StructureBlockInfo> processedBlockInfos, StructurePlaceSettings settings) {
+        //Map<BlockPos, StructureTemplate.StructureBlockInfo> mapByPos = mapByPos(processedBlockInfos);
+        List<StructureTemplate.StructureBlockInfo> newBlockInfos = new ArrayList<>();
+        for (StructureTemplate.StructureBlockInfo blockInfo : processedBlockInfos) {
+            StructureTemplate.StructureBlockInfo newBlockInfo = processFinal(serverLevel, offset, pos, blockInfo, blockInfo, settings, processedBlockInfos);
+            newBlockInfos.add(newBlockInfo);
+        }
+        return newBlockInfos;
+    }
+
+
+    public StructureTemplate.StructureBlockInfo processFinal(LevelReader world, BlockPos piecePos, BlockPos structurePos, StructureTemplate.StructureBlockInfo rawBlockInfo, StructureTemplate.StructureBlockInfo blockInfo, StructurePlaceSettings settings, List<StructureTemplate.StructureBlockInfo> processedBlockInfos) {
+        RandomSource random = ProcessorUtil.getRandom(structureRandomType, blockInfo.pos(), piecePos, structurePos, world, seed);
+        BlockState blockstate = blockInfo.state();
+        BlockPos blockpos = blockInfo.pos();
+        if(blockstate.isAir() && random.nextFloat() <= rarity){
+            boolean hasSides = isHasSides(blockpos, processedBlockInfos);
+            boolean hasUp = !requiresUp || hasDirection(processedBlockInfos, rawBlockInfo.pos(), Direction.UP);
+            boolean hasDown = !requiresDown || hasDirection(processedBlockInfos, rawBlockInfo.pos(), Direction.DOWN);
+            if(hasSides && hasUp && hasDown){
+                return new StructureTemplate.StructureBlockInfo(blockpos, blockState, blockInfo.nbt());
+            }
+        }
+        return blockInfo;
+    }
+
+    private boolean isHasSides(BlockPos blockpos, List<StructureTemplate.StructureBlockInfo> processedBlockInfos) {
+        if(requiresSides == 0) return true;
+        int sides = 0;
+        if(hasDirection(processedBlockInfos, blockpos, NORTH)) {
```

Comment:
We should probably put an ImmutableList of horizontal directions in a util class. Then you could iterate over it.